tests/unit/code_provider/test_user_local_tunnel_provider.py

```
@@ -0,0 +1,703 @@
+
+    @patch.object(UserLocalTunnelProvider, "_make_tunnel_request")
+    def test_success_false_raises_file_not_found_error(self, mock_tunnel_request):
+        """Test that success:False raises FileNotFoundError."""
+        mock_tunnel_request.return_value = {
+            "success": False,
+            "error": "File not found",
+        }
+
+        provider = UserLocalTunnelProvider(tunnel_url="https://tunnel.example.com")
+        with pytest.raises(
+            FileNotFoundError, match="Failed to read file 'test.py': File not found"
```

Ruff (0.15.10)

[warning] Pattern passed to `match=` contains metacharacters but is neither escaped nor raw

(RUF043)